neuronunit/models/backends.py

```diff
@@ -423,28 +402,127 @@ def inject_square_current(self, current):
+
+        assert type(self.model.attrs) is not type(None)
+        for h_key,h_value in attrs.items():
+            self.h('m_RS_RS_pop[0].{0} = {1}'.format(h_key,h_value))
+            self.h('m_{0}_{1}_pop[0].{2} = {3}'.format(self.cell_name,self.cell_name,h_key,h_value))
+
+        # Below are experimental rig recording parameters.
+        # These can possibly go in a seperate method.
+
+        self.h(' { v_time = new Vector() } ')
+        self.h(' { v_time.record(&t) } ')
+        self.h(' { v_v_of0 = new Vector() } ')
+        self.h(' { v_v_of0.record(&RS_pop[0].v(0.5)) } ')
+        self.h(' { v_u_of0 = new Vector() } ')
+        self.h(' { v_u_of0.record(&m_RS_RS_pop[0].u) } ')
+
+        self.tVector = self.h.v_time
+        self.vVector = self.h.v_v_of0
+        '''
+        ##
+        # Empty the cache when redefining the model
+        ##
+        self.model.cached_params = {}
+        self.model.cached_attrs = {}
+        self.current = {}
+        #print('the cache emptied: {0}'.format(str(self.model.cached_params)))
+        #print('As the attributes are updated: {0}'.format(str(self.model.attrs)))
+        return self
+
+    def inject_square_current(self, current):
+        def dict_hash(d):
+            return hash(pickle.dumps([(key,x[key]) for key in sorted(d)]))
+
+
+        if str(self.model.attrs) not in self.cached_attrs:
+            results = super(MemoryBackend,self).local_run()#
+            self.model.cached_attrs[dict_hash(self.model.attrs)] = 1
+        else:
+            self.model.cached_attrs[dict_hash(self.model.attrs)] += 1
+
+        super(MemoryBackend,self).inject_square_current(current)#
+        #
+        # make this current injection value a class attribute, such that its remembered.
+        #
+        #from sciunit.utils import dict_hash
+
+        if 'injected_square_current' in current:
+            self.current = current['injected_square_current']
+        else:
+            self.current = current
+        if str(self.current) not in self.model.cached_params:
+
+            results = super(MemoryBackend,self).local_run()#
+            self.model.cached_params[str(self.current)] = {}
+            self.model.cached_params[str(self.current)]=results
+            #print('the cache: {0}'.format(str(self.model.cached_params.keys())))
+
+        return self.model.cached_params[str(self.current)]
+
+
+    def local_run(self):
+        #
+        # Logic for choosing whether a injection value, has already been tested.
+        #
+        if str(self.current) not in self.model.cached_params:
+            results = super(MemoryBackend,self).local_run()
+            self.model.cached_params[str(self.current)] = {}
+            self.model.cached_params[str(self.current)]=results
+            #print('the cache: {0}'.format(str(self.model.cached_params)))
+
+            return self.model.cached_params[str(self.current)]
+        else:
+            return self.model.cached_params[str(self.current)]
+
```

Comment:
@russelljjarvis 
Having already used `NEURONMemoryBackend` to examine the issue you were looking into (about whether the optimizer checks the same set of parameters twice), could you remove it so we can let the more general caching implementation take over?